File: Tests/BE/business/case/Obligation.py
from ..base.Base import Base
from .ObligationCreditor import ObligationCreditor
from .ObligationCorrection import ObligationCorrection

from common import Convertor

from typing import List

DISPATCHES = {
            'obligation_type_id': lambda value: Convertor.to_int(value),
            'obligation_state': lambda value: Convertor.to_int(value),
            'installment_amount': lambda value: Convertor.to_decimal(value),
            'loan_principal_amount': lambda value: Convertor.to_decimal(value),
            'credit_card_limit': lambda value: Convertor.to_decimal(value),
            'amount_consolidated': lambda value: Convertor.to_decimal(value),

            'creditor': lambda value: ObligationCreditor.from_json(value),
            'correction': lambda value: ObligationCorrection.from_json(value),
        }

JSON_KEYS = ['obligationTypeId', 'obligationState', 'installmentAmount', 'loanPrincipalAmount', 'creditCardLimit', 'amountConsolidated', 'creditor', 'correction']

class Obligation(Base):

    def __init__(self, js_dict: dict = None):
        super().__init__(dispatches= DISPATCHES , js_dict = js_dict)

    @staticmethod
    def from_json(js_dict: dict):
        return Obligation(js_dict = js_dict)

    @staticmethod
    def from_json_list(js_dict_list: List[dict]):
        return list(map(lambda i: Obligation.from_json(i), js_dict_list))

    @staticmethod
    def to_grpc(value: object):
        if (value is None):
            return None
        assert isinstance(value, Obligation)
        return value.to_grpc()

    def to_grpc(self) -> dict:
       return dict()

    def _get_json_keys(self):
        return JSON_KEYS


# ?????:
# obligationId
# scustomerOnSAId

# --------------------------------------------------------------------------------------------
#                 "obligations": [
#                     {
#                         "obligationId": 43,
#                         "customerOnSAId": 291,
#                         "obligationTypeId": 1,
#                         "installmentAmount": 10000,
#                         "loanPrincipalAmount": 1000000,
#                         "creditor": {
#                             "creditorId": "",
#                             "name": "",
#                             "isExternal": true
#                         },
#                         "correction": {
#                             "correctionTypeId": 1
#                         }
#                     }
#                 ]
# --------------------------------------------------------------------------------------------

File: DomainServices/HouseholdService/Api/Database/DocumentDataEntities/Obligation.cs
namespace DomainServices.HouseholdService.Api.Database.DocumentDataEntities;

internal sealed class Obligation
    : SharedComponents.DocumentDataStorage.IDocumentData
{
    public int Version => 1;

    public int? ObligationTypeId { get; set; }
    public decimal? InstallmentAmount { get; set; }
    public decimal? LoanPrincipalAmount { get; set; }
    public decimal? CreditCardLimit { get; set; }
    public decimal? AmountConsolidated { get; set; }
    public int? ObligationState { get; set; }
    public ObligationCreditor? Creditor { get; set; }
    public ObligationCorrection? Correction { get; set; }

    internal sealed class ObligationCreditor
    {
        public string? CreditorId { get; set; }
        public string? Name { get; set; }
        public bool? IsExternal { get; set; }
    }

    internal sealed class ObligationCorrection
    {
        public int? CorrectionTypeId { get; set; }
        public decimal? InstallmentAmountCorrection { get; set; }
        public decimal? LoanPrincipalAmountCorrection { get; set; }
        public decimal? CreditCardLimitCorrection { get; set; }
    }
}
